File: new/group_anagrams_hash_table.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // create hash table to store values of stringed<frequency counter> : list of strings
        // we create a vector frequency counter to count each char for each string, and string this frequency counter. because vector is not hashable in c++ we need something hashable to store as keys in our map. 
        // for each key, store relevant list of anagrams of strings as list
        // return the list of anagrams for each key as master list. 
        unordered_map<string, vector<string>> map;
        for (const auto &s: strs) {
            vector<int> frequency_counter(26, 0);
            for (auto c: s) {
                frequency_counter[c - 'a'] += 1;
            }

            // because vector is not hashable, we need to have something hashable for keys in a map
            string frequency_string = to_string(frequency_counter[0]);
            for (int i = 1; i < 26; i++) {
                frequency_string += "," + to_string(frequency_counter[i]);
            }

            map[frequency_string].push_back(s);
        }

        vector<vector<string>> result;
        for (const auto &pair: map) {
            result.push_back(pair.second);
        }

        return result;
    }
    // time: O(n * m) - n for each char in each string, m for each string in the original list
    // space: O(n * m) ~= O(m) because n is always 26 in alphabets. 
};

File: new/python/group_anagrams_hash_table.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # create a hashmap to contain tuple: list of strings
        # we create the list then convert it to tuple when we use to index and append values
        # tuple is ideal because we are going to create a list of 26 positions for 26 chars
        # this tuple is like a frequency map
        # for each tuple we will append relevant string from origianl list to it
        # then return the list of values of the map
        map = defaultdict(list)  # we need to declare type of values for default dict
        for s in strs:
            char_frequency = [0] * 26
            for c in s:
                char_frequency[ord(c) - ord("a")] += 1

            tuple_char_frequency = tuple(
                char_frequency
            )  # because tuple is hashable in python and list not hashable in python
            map[tuple_char_frequency].append(s)

        return list(map.values())
        # time: O(m * n) because we are not sorting anything so no nlogn, we go through m strings in list, and for each string we go through n characters
        # space: O(m * n) ~= O(m) because we have constant space for the 26 character frequency list worst case we have m number of keys in the map and each one is unique.
